agave_app/renderer.cpp

```diff
@@ -165,76 +273,92 @@ Renderer::processRequest()
       // push another redraw request.
       std::vector<Command*> cmd;
       RequestRedrawCommandD data;
       cmd.push_back(new RequestRedrawCommand(data));
-      this->addRequest(new RenderRequest(lastReq->getClient(), cmd, false));
-    }
+      RenderRequest* rr = new RenderRequest(lastReq->getClient(), cmd, false);
 
-    // inform the server that we are done with r
-    emit requestProcessed(lastReq, img);
+      // this->addRequest(rr);
+
+      this->m_requests << rr;
+      this->m_totalQueueDuration += rr->getDuration();
+    }
 
   } else {
     // if not in stream mode, then process one request, then re-render.
+    if (!this->m_requests.isEmpty() && !this->isInterruptionRequested()) {
 
-    // remove request from the queue
-    RenderRequest* r = this->m_requests.takeFirst();
-    this->m_totalQueueDuration -= r->getDuration();
-
-    // process it
-    QElapsedTimer timer;
-    timer.start();
+      // remove request from the queue
+      RenderRequest* r = this->m_requests.takeFirst();
+      this->m_totalQueueDuration -= r->getDuration();
 
-    std::vector<Command*> cmds = r->getParameters();
-    if (cmds.size() > 0) {
-      this->processCommandBuffer(r);
-    }
+      // process it
+      QElapsedTimer timer;
+      timer.start();
 
-    QImage img = this->render();
+      std::vector<Command*> cmds = r->getParameters();
+      if (cmds.size() > 0) {
+        this->processCommandBuffer(r);
+      }
 
-    r->setActualDuration(timer.nsecsElapsed());
+      img = this->render();
 
-    // inform the server that we are done with r
-    emit requestProcessed(r, img);
+      r->setActualDuration(timer.nsecsElapsed());
+      lastReq = r;
+    }
   }
 
+  // unlock mutex BEFORE emit, in case the signal handler wants to add a request
+  m_requestMutex.unlock();
+
+  // inform the server that we are done with r
+  // TODO : have a mode where we don't need a QImage
+  // and can just return rgba byte array as a thread safe shared ptr
+  // writable by render thread and readable by anyone else
+  if (!this->isInterruptionRequested()) {
+    emit requestProcessed(lastReq, img);
+  }
   return true;
 }
 
 void
 Renderer::processCommandBuffer(RenderRequest* rr)
 {
-#if HAS_EGL
-  this->m_glContext->makeCurrent();
-#else
-  this->m_glContext->makeCurrent(this->m_surface);
-#endif
+  m_rglContext.makeCurrent();
 
   std::vector<Command*> cmds = rr->getParameters();
   if (cmds.size() > 0) {
-    m_ec.m_renderSettings = m_myVolumeData.m_renderSettings;
+    // get the renderer we need
+    RenderGLPT* r = dynamic_cast<RenderGLPT*>(m_myVolumeData.m_renderer);
+    if (!r) {
+      LOG_ERROR << "Unsupported renderer: Renderer is not of type RenderGLPT";
+    }
+
+    m_ec.m_renderSettings = &r->getRenderSettings(); // m_myVolumeData.m_renderSettings;
     m_ec.m_renderer = this;
-    m_ec.m_appScene = m_myVolumeData.m_scene;
+    m_ec.m_appScene = r->scene(); // m_myVolumeData.m_scene;
```

Comment:
Will these commented-out assignments be useful for understanding or testing later?

Reply:
no, just cruft -- I'll remove.